File: MIDTERM-LAB-3/MP20/MP20.java
package MP20;
/*
 * =====================================================
 * Student Name    : MACALAM, CAMRY S.
 * Course          : Programming 2
 * Assignment      : Machine Problem 20 — Convert CSV Dataset to JSON
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 16, 2026
 * GitHub Repo     : https://github.com/macalam-camry/uphsd-cs-macalam-camry
 *
 * Description:
 *   Reads the CSV dataset from a user-supplied file path, parses each
 *   data record using the column header row as property keys, and
 *   builds a JSON array where every element is one record object.
 *   The resulting JSON is written to dataset.json and a preview of
 *   the first five objects is printed to the console so the output
 *   can be verified at a glance without opening the file.
 * =====================================================
 */

import java.io.*;
import java.util.*;

public class MP20 {

    // Column header names we assign to each meaningful field.
    // The original header row is used for indexes 0,1; index 2 is the
    // always-empty spacer which we skip; 3–8 are the remaining data fields.
    static final String[] FIELD_NAMES = {
        "candidate", "type", "exam", "language",
        "examDate", "score", "result", "timeUsed"
    };

    // Column indices from the raw CSV that map to the above field names.
    // Index 2 (Column1 / spacer) is deliberately omitted.
    static final int[] FIELD_COLS = { 0, 1, 3, 4, 5, 6, 7, 8 };

    // Number of metadata/title lines before the real column header row
    static final int METADATA_LINES = 6;

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        // ── Step 1: Prompt for file path and validate ────────────────────
        File file = null;
        while (true) {
            System.out.print("Enter dataset file path: ");
            String filePath = input.nextLine().trim();
            file = new File(filePath);

            if (!file.exists() || !file.isFile()) {
                System.out.println("Error: File not found. Please try again.\n");
            } else if (!file.canRead()) {
                System.out.println("Error: File is not readable. Please try again.\n");
            } else if (!filePath.toLowerCase().endsWith(".csv")) {
                System.out.println("Error: Not a CSV file. Please try again.\n");
            } else {
                System.out.println("File found. Processing...\n");
                break;
            }
        }

        // ── Step 2: Parse CSV into a list of field-value maps ────────────
        // Each entry in the list represents one record as a LinkedHashMap
        // so the field insertion order is preserved in the JSON output.
        List<Map<String, String>> records = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line;
            int lineIndex = 0;

            while ((line = br.readLine()) != null) {
                // Skip all metadata rows and the column header row
                if (lineIndex <= METADATA_LINES) { lineIndex++; continue; }

                String[] cols = parseCSVLine(line);

                // Skip completely blank rows
                boolean allBlank = true;
                for (String c : cols) { if (!c.trim().isEmpty()) { allBlank = false; break; } }
                if (allBlank) { lineIndex++; continue; }

                // Build a key-value map for this record using our field names
                Map<String, String> record = new LinkedHashMap<>();
                for (int i = 0; i < FIELD_COLS.length; i++) {
                    int colIdx = FIELD_COLS[i];
                    String val = colIdx < cols.length ? cols[colIdx].trim() : "";
                    record.put(FIELD_NAMES[i], val);
                }
                records.add(record);
                lineIndex++;
            }
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            input.close();
            return;
        }

        if (records.isEmpty()) {
            System.out.println("No records found.");
            input.close();
            return;
        }

        // ── Step 3: Build the JSON string ────────────────────────────────
        // We construct the JSON manually without any external library,
        // which satisfies the 'no math/IO libraries for core logic' rule.
        StringBuilder json = new StringBuilder();
        json.append("[\n");

        for (int i = 0; i < records.size(); i++) {
            Map<String, String> record = records.get(i);
            json.append("  {\n");
            int fieldIdx = 0;
            for (Map.Entry<String, String> entry : record.entrySet()) {
                json.append("    \"")
                    .append(jsonEscape(entry.getKey()))
                    .append("\": \"")
                    .append(jsonEscape(entry.getValue()))
                    .append("\"");
                // Add a comma after every field except the last one
                if (fieldIdx < record.size() - 1) json.append(",");
                json.append("\n");
                fieldIdx++;
            }
            // Comma after every object except the final one
            json.append("  }");
            if (i < records.size() - 1) json.append(",");
            json.append("\n");
        }
        json.append("]");

        // ── Step 4: Write JSON to file ────────────────────────────────────
        String outputFile = "dataset.json";
        try (FileWriter fw = new FileWriter(outputFile)) {
            fw.write(json.toString());
        } catch (IOException e) {
            System.out.println("Error writing JSON file: " + e.getMessage());
            input.close();
            return;
        }

        // ── Step 5: Print confirmation and a 5-record preview ────────────
        System.out.println("=".repeat(70));
        System.out.println("  MP20 — CSV TO JSON CONVERSION");
        System.out.println("  Student: MACALAM, CAMRY S.");
        System.out.println("=".repeat(70));
        System.out.printf("  Records converted : %d%n",   records.size());
        System.out.printf("  Output file       : %s%n",   outputFile);
        System.out.println("=".repeat(70));
        System.out.println("\n  PREVIEW — First 5 JSON Objects:");
        System.out.println("  " + "-".repeat(68));

        int previewCount = Math.min(5, records.size());
        System.out.println("  [");
        for (int i = 0; i < previewCount; i++) {
            Map<String, String> record = records.get(i);
            System.out.println("    {");
            int fi = 0;
            for (Map.Entry<String, String> entry : record.entrySet()) {
                String comma = fi < record.size() - 1 ? "," : "";
                System.out.printf("      \"%s\": \"%s\"%s%n",
                        entry.getKey(), entry.getValue(), comma);
                fi++;
            }
            String objComma = i < previewCount - 1 ? "," : "";
            System.out.println("    }" + objComma);
        }
        if (records.size() > 5) {
            System.out.printf("    ... and %d more records%n", records.size() - 5);
        }
        System.out.println("  ]");
        System.out.println("=".repeat(70));

        input.close();
    }

    // ── Escape special characters for JSON string values ─────────────────
    static String jsonEscape(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    // ── Parse a single CSV line, respecting quoted fields ────────────────
    static String[] parseCSVLine(String line) {
        List<String> result = new ArrayList<>();
        boolean inQuotes = false;
        StringBuilder current = new StringBuilder();
        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                result.add(current.toString());
                current.setLength(0);
            } else {
                current.append(c);
            }
        }
        result.add(current.toString());
        return result.toArray(new String[0]);
    }
}

File: MIDTERM-LAB-3/MP20/mp20.js
/**
 * =====================================================
 * Student Name    : MACALAM, CAMRY S.
 * Course          : Programming 2
 * Assignment      : Machine Problem 20 — Convert CSV Dataset to JSON
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 16, 2026
 * GitHub Repo     : https://github.com/macalam-camry/uphsd-cs-macalam-camry
 * Runtime         : Node.js — run with: node mp20.js
 *
 * Description:
 *   Reads the CSV dataset from a user-supplied file path, parses each
 *   data record using the column header row as property keys, and
 *   builds a JSON array where every element is one record object.
 *   The resulting JSON is written to dataset.json and a preview of
 *   the first five objects is printed to the console so the output
 *   can be verified at a glance without opening the file.
 * =====================================================
 */

const fs       = require('fs');
const readline = require('readline');
const path     = require('path');

// Human-readable property names mapped to each meaningful CSV column index.
// The always-empty spacer at index 2 is deliberately excluded.
const FIELD_MAP = [
    { name: 'candidate', col: 0 },
    { name: 'type',      col: 1 },
    { name: 'exam',      col: 3 },
    { name: 'language',  col: 4 },
    { name: 'examDate',  col: 5 },
    { name: 'score',     col: 6 },
    { name: 'result',    col: 7 },
    { name: 'timeUsed',  col: 8 }
];

// Number of metadata/title lines before the real column header row
const METADATA_LINES = 6;

// ── Parse a single CSV line, respecting quoted fields ────────────────
function parseCSVLine(line) {
    const result = [];
    let inQuotes = false;
    let current  = '';
    for (const c of line) {
        if (c === '"') {
            inQuotes = !inQuotes;
        } else if (c === ',' && !inQuotes) {
            result.push(current);
            current = '';
        } else {
            current += c;
        }
    }
    result.push(current);
    return result;
}

// ── Load the CSV and convert every data row into a plain JS object ───
function processFile(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');
    const lines   = content.split(/\r?\n/);

    // Build the array of record objects
    const records = [];

    for (let i = 0; i < lines.length; i++) {
        if (i <= METADATA_LINES) continue;          // skip metadata + header

        const cols = parseCSVLine(lines[i]);
        if (cols.every(c => c.trim() === '')) continue; // skip blank rows

        // Map each meaningful column to its friendly property name
        const record = {};
        for (const { name, col } of FIELD_MAP) {
            record[name] = col < cols.length ? cols[col].trim() : '';
        }
        records.push(record);
    }

    if (records.length === 0) {
        console.log('No records found.');
        return;
    }

    // ── Serialise to JSON using built-in JSON.stringify ───────────────
    // The third argument (2) adds two-space indentation for readability.
    const jsonOutput = JSON.stringify(records, null, 2);

    // ── Write the JSON file ───────────────────────────────────────────
    const outputFile = 'dataset.json';
    fs.writeFileSync(outputFile, jsonOutput, 'utf8');

    // ── Print confirmation and a 5-record preview ─────────────────────
    const SEP  = '='.repeat(70);
    const DASH = '-'.repeat(68);

    console.log(SEP);
    console.log('  MP20 — CSV TO JSON CONVERSION');
    console.log('  Student: MACALAM, CAMRY S.');
    console.log(SEP);
    console.log(`  Records converted : ${records.length}`);
    console.log(`  Output file       : ${outputFile}`);
    console.log(SEP);
    console.log('\n  PREVIEW — First 5 JSON Objects:');
    console.log('  ' + DASH);

    // Show just the first 5 records in pretty-printed form
    const preview = records.slice(0, 5);
    console.log('  [');
    preview.forEach((rec, i) => {
        console.log('    {');
        const keys = Object.keys(rec);
        keys.forEach((key, ki) => {
            const comma = ki < keys.length - 1 ? ',' : '';
            console.log(`      "${key}": "${rec[key]}"${comma}`);
        });
        const objComma = i < preview.length - 1 ? ',' : '';
        console.log('    }' + objComma);
    });
    if (records.length > 5) {
        console.log(`    ... and ${records.length - 5} more records`);
    }
    console.log('  ]');
    console.log(SEP);
}

// ── Main: prompt for file path, validate, then run ───────────────────
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

function askFilePath() {
    rl.question('Enter dataset file path: ', inputPath => {
        inputPath = inputPath.trim();

        if (!fs.existsSync(inputPath)) {
            console.log('Error: File not found. Please try again.\n');
            return askFilePath();
        }
        try { fs.accessSync(inputPath, fs.constants.R_OK); }
        catch { console.log('Error: File is not readable. Please try again.\n'); return askFilePath(); }
        if (path.extname(inputPath).toLowerCase() !== '.csv') {
            console.log('Error: Not a CSV file. Please try again.\n');
            return askFilePath();
        }

        console.log('File found. Processing...\n');
        try {
            processFile(inputPath);
        } catch (err) {
            console.log('Error processing file: ' + err.message);
        }
        rl.close();
    });
}

askFilePath();
